bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -2525,185 +2570,206 @@ void BCStateTran::processData(bool lastInBatch) {
+            finalizePutblockAsync(PutBlockWaitPolicy::WAIT_SINGLE_JOB);
+          }
+          auto ctx = ioPool_.alloc();
+          ctx->blockId = nextRequiredBlock_;
+          ctx->actualBlockSize = actualBlockSize;
+          // TODO - this can probably be optimized - see TODO above getNextFullBlock
+          memcpy(ctx->blockData.get(), buffer_.get(), actualBlockSize);
+          ctx->future = as_->putBlockAsync(nextRequiredBlock_, ctx->blockData.get(), ctx->actualBlockSize, false);
+          ioContexts_.push_back(std::move(ctx));
+          histograms_.dst_num_pending_blocks_to_commit->record(ioContexts_.size());
+          finalizePutblockAsync(PutBlockWaitPolicy::NO_WAIT);
+          as_->getPrevDigestFromBlock(
+              buffer_.get(), actualBlockSize, reinterpret_cast<StateTransferDigest *>(&digestOfNextRequiredBlock_));
+          ConcordAssertGT(nextRequiredBlock_, 0);
+          --nextRequiredBlock_;
+          LOG_TRACE(logger_, KVLOG(nextRequiredBlock_));
+          if (lastInBatch || posponedSendFetchBlocksMsg_ || newSourceReplica) {
+            trySendFetchBlocksMsg(firstRequiredBlock,
+                                  nextRequiredBlock_,
+                                  0,
+                                  KVLOG(lastInBatch, posponedSendFetchBlocksMsg_, newSourceReplica));
+            break;
+          }
+        } else {
+          //////////////////////////////////////////////////////////////////////////
+          // This is the last block we need in cycle or on a fetch range
+          //////////////////////////////////////////////////////////////////////////
+          LOG_INFO(logger_, ss.str());
+          commitToChainDT_.start();
+          blocks_collected_.pause();
+          bytes_collected_.pause();
+          finalizePutblockAsync(PutBlockWaitPolicy::WAIT_ALL_JOBS, true);
+
+          ConcordAssertEQ(nextBlockIdToCommit_, nextRequiredBlock_);
+          ConcordAssert(ioContexts_.empty());
+          ConcordAssert(as_->putBlock(nextRequiredBlock_, buffer_.get(), actualBlockSize, true));
+
+          commitToChainDT_.pause();
+          digestOfNextRequiredBlock_.makeZero();
+          clearAllPendingItemsData();
+          DataStoreTransaction::Guard g(psd_->beginTransaction());
+          {
+            if (!lastFetchedBlockIdInCycle) {
+              const auto &lastSentMsg = std::get<FetchBlocksMsg>(lastSentMsg_);
+              ConcordAssertLE(lastSentMsg.maxBlockId + 1, psd_->getLastRequiredBlock());
+              g.txn()->setFirstRequiredBlock(lastSentMsg.maxBlockId + 1);
+              nextRequiredBlock_ = computeNextRequiredBlock();
+              nextBlockIdToCommit_ = nextRequiredBlock_;
+              trySendFetchBlocksMsg(psd_->getFirstRequiredBlock(),
+                                    nextRequiredBlock_,
+                                    0,
+                                    KVLOG(lastInBatch, posponedSendFetchBlocksMsg_, newSourceReplica));
```

Comment:
nit:
Should posponedSendFetchBlocksMsg_ be corrected?

Reply:
Thanks, I did. Added a commit.
not sure if this one works anymore.
Will have to validate after decoupling.